### DISCUSSION TITLE
IG Video Downloader

### DISCUSSION THREAD

**Discussion by ashutoshkrris** [ashutoshkrris/EazyLoader#16] [General]
Will the IG Video Downloader work if the project is deployed on Heroku?

**Comment by Anubhab2002** [comment]
Yeah it works, only thing is, during deployment you need to add two buildpacks:
```$ herokubuildpacks:add --index 1 https://github.com/heroku/heroku-buildpack-chromedriver```
```$ heroku buildpacks:add --index 2 https://github.com/heroku/heroku-buildpack-google-chrome```
as mentioned [here](https://elements.heroku.com/buildpacks/heroku/heroku-buildpack-chromedriver#:~:text=This%20buildpack%20installs%20chromedriver%20(the,you'll%20also%20need%20Chrome. )

checkout [this](https://stackoverflow.com/questions/42860703/how-to-use-selenium-webdriver-on-heroku) for details.

**Comment by Anubhab2002** [comment]
This medium blog is a good one, specifically for python projects: [deploy python selenium project on heroku](https://medium.com/@mikelcbrowne/running-chromedriver-with-python-selenium-on-heroku-acc1566d161c)

**Comment by ashutoshkrris** [comment]
Thanks a bunch! @Anubhab2002 
Will follow the same
If I encounter some bugs, will discuss here

**Reply by Anubhab2002** [reply]
sure thing!

**Comment by Anubhab2002** [comment]
you are welcome @ashutoshkrris!!
By the way, you could now move instagram video downloader from upcoming features section to the features section in README.md

**Reply by ashutoshkrris** [reply]
Ahh Yes!
Forgot about that, thanks

**Comment by ashutoshkrris** [comment]
@Anubhab2002 
If I keep `chrome_options.headless=True`, it's not able to fetch the video.
Is there any solution for that, as we neee to put --headless flag on Heroku

**Reply by Anubhab2002** [reply]
I don't think --headless argument is necessary for deployment, only --no-sandbox and --disable-gpu are needed

**Reply by ashutoshkrris** [reply]
If remove the --headless argument, I get this error: 
```bash
Message: unknown error: Chrome failed to start: exited abnormally.
2021-10-07T09:11:05.659039+00:00 app[web.1]:   (unknown error: DevToolsActivePort file doesn't exist)
2021-10-07T09:11:05.659040+00:00 app[web.1]:   (The process started from chrome location /app/.apt/opt/google/chrome/chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
```

**Reply by Anubhab2002** [reply]
Just confirming, You added the buildpacks right?

**Reply by ashutoshkrris** [reply]
Yep, those things are set up correctly.
The line which is posing the problem is : 
```python
WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.CLASS_NAME, "_5wCQW")))
```

Looks like this doesn't work with headless chrome

**Reply by Anubhab2002** [reply]
yeah basically it is not allowing the pop-up screen to open for the scrapping to take place...anyways wait a bit. Let me try I can can refactor it a bit.

**Reply by ashutoshkrris** [reply]
Tried taking screenshot of the page in headless mode. This is what I got :

![ss](https://user-images.githubusercontent.com/47353498/136362179-6dc5ad87-fcab-4f08-9290-66e3923d71e5.png)

**Reply by Anubhab2002** [reply]
So is the issue fixed now?

**Comment by ashutoshkrris** [comment]
Nope
Also, I found..on local environment, it's not asking to login
But on heroku, Instagram asks to login

**Comment by Anubhab2002** [comment]
Hey @ashutoshkrris, it worked just fine on my local
![Screenshot from 2021-10-07 18-20-20](https://user-images.githubusercontent.com/84004920/136387417-33c7116c-d170-421d-badc-a2cbea30b1cb.png)

**Reply by ashutoshkrris** [reply]
Yeah, that is working fine
But, on the deployed project, it's asking to login

**Reply by Anubhab2002** [reply]
but after logging in are we getting our job done?

**Reply by ashutoshkrris** [reply]
Most probably, it should
But I am not able to login programmatically

I am using this code : 
```python
if 'login' in driver.current_url:
    driver.implicitly_wait(20)
    driver.find_element_by_xpath('//*[@id="loginForm"]/div/div[1]/div/label/input').send_keys('username')
    password=driver.find_element_by_xpath('//*[@id="loginForm"]/div/div[2]/div/label/input')
    password.send_keys('pwd')
    password.send_keys(Keys.ENTER)
```

Error : 
```bash
Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="loginForm"]/div/div[1]/div/label/input"}
```

**Reply by Anubhab2002** [reply]
your code is alright, instead of implicit wait, try adding a sleep() before each of username, password, click. like maybe a sleep(5) before each action is performed

**Reply by Anubhab2002** [reply]
this issue arises can arise when the element has not loaded yet and you are trying to access it, so maybe try increasing the sleep time

**Comment by Anubhab2002** [comment]
@ashutoshkrris working now?

**Reply by ashutoshkrris** [reply]
No, I tried with this code, still failing

```python
            driver.get(url)
            print("Page opened")
            print(driver.current_url)
            time.sleep(10)
            if 'login' in driver.current_url:
                driver.find_element_by_name(
                    'username').send_keys(IG_USERNAME)
                time.sleep(2)
                password = driver.find_element_by_name(
                    'password')
                time.sleep(2)
                password.send_keys(IG_PASSWORD)
                password.send_keys(Keys.ENTER)
                driver.get(url)
```

The first driver.get(url) can lead us to login page. So after logging in, we are using driver.get(url) again to get the correct page. But its opening the login page

**Reply by Anubhab2002** [reply]
I see. Well this is going beyond my understanding now. Best is to try asking on Stack Overflow.

**Comment by Anubhab2002** [comment]
@ashutoshkrris can stopping **redirect** in ```driver.get(url)``` and specifying the acceptable forms of the url help in this case?

**Reply by ashutoshkrris** [reply]
Sorry, didn't get you

**Reply by Anubhab2002** [reply]
as far as I understood the problem, when you are putting the url in the ig video downloader, the link is redirecting you to the login page. So in that case if we can somehow stop the redirect then can we get the viable results?

**Reply by ashutoshkrris** [reply]
We cannot do it using Selenium, i suppose

**Comment by Anubhab2002** [comment]
@ashutoshkrris , what is the present scenario?

**Comment by ashutoshkrris** [comment]
Couldn't figure out anything, so I have put the feature on hold.

**Reply by Anubhab2002** [reply]
I see. I am sorry, I tried as much as I could. Still I will look for alternatives.

**Reply by ashutoshkrris** [reply]
Yeah, I'm also trying.
I suppose Instagram is very strict regarding headless bots.
